flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/agglomerativeclustering/AgglomerativeClustering.java

```diff
@@ -0,0 +1,415 @@
+        }
+
+        private void doClustering(List<Cluster> activeClusters) {
+            int clusterOffset1 = -1, clusterOffset2 = -1;
+            boolean clusteringRunning =
+                    (k != null && activeClusters.size() > k) || (distanceThreshold != null);
+
+            while (clusteringRunning || (computeFullTree && activeClusters.size() > 1)) {
+                // Computes the distance between two clusters.
+                double minDistance = Double.MAX_VALUE;
+                for (int i = 0; i < activeClusters.size(); i++) {
+                    for (int j = i + 1; j < activeClusters.size(); j++) {
+                        double distance =
+                                computeDistanceBetweenClusters(
+                                        activeClusters.get(i), activeClusters.get(j));
+                        if (distance < minDistance) {
+                            minDistance = distance;
+                            clusterOffset1 = i;
+                            clusterOffset2 = j;
+                        }
+                    }
+                }
+
+                // Outputs the merge info.
+                Cluster cluster1 = activeClusters.get(clusterOffset1);
+                Cluster cluster2 = activeClusters.get(clusterOffset2);
+                int clusterId1 = cluster1.clusterId;
+                int clusterId2 = cluster2.clusterId;
+                output.collect(
+                        mergeInfoOutputTag,
+                        new StreamRecord<>(
+                                Tuple4.of(
+                                        Math.min(clusterId1, clusterId2),
+                                        Math.max(clusterId1, clusterId2),
+                                        minDistance,
+                                        cluster1.dataPointIds.size()
+                                                + cluster2.dataPointIds.size())));
+
+                // Merges these two clusters.
+                Cluster mergedCluster =
+                        new Cluster(
+                                getNextClusterId(), cluster1.dataPointIds, cluster2.dataPointIds);
+                activeClusters.set(clusterOffset1, mergedCluster);
+                activeClusters.remove(clusterOffset2);
+
+                // Updates cluster Ids for each data point if clustering is still running.
+                if (clusteringRunning) {
+                    int mergedClusterId = mergedCluster.clusterId;
+                    for (int dataPointId : mergedCluster.dataPointIds) {
+                        clusterIds[dataPointId] = mergedClusterId;
+                    }
+                }
+
+                clusteringRunning =
+                        (k != null && activeClusters.size() > k)
+                                || (distanceThreshold != null && distanceThreshold > minDistance);
+            }
+        }
+
+        private double computeDistanceBetweenClusters(Cluster cluster1, Cluster cluster2) {
```

Comment:
How about adding a cache here for the distance between any two data points? We can avoid re-computing the distance if there is a cache hit.

Reply:
The memory overhead for caching the distance between two data points is O(n^2), where n is the number of data points.

Suppose we store the distance in a primitive double array, then the max number of data points we can deal with is Math.sqrt(Int.MAX), which is less than 100000. 

A possible optimization would be cache the distance between each two clusters at the later iterations, i.e., when the number of active clusters is smaller. But we can do this in a later PR.